### PR TITLE
Check if snapcraft.yaml is present

### LINKED ISSUE #2449: Check whether snapcraft.yaml is present and show a message if not
on page load
Are there builds?

YES: 
 - load the page
 
NO: 
 - check against github repo (see build for possible snapcraft.yaml locations and names)
 - if now present - trigger the first build
 - if not present - do nothing

### PR DESCRIPTION
**Pull request by jkfran**
## Done

In this PR we are using the GitHub GraphQL API to:

- List all the user repositories and mark the ones with a snapcraft.yaml present:
github-graphql1

- Show an error on the builds page when the selected repo for the snap doesnt contain a snapcraft.yaml:
github-graphql2

## Issue / Card

Fixes #2449

## QA

- Pull the branch
- Run the site using the command `./run`
- View the site locally in your web browser at: http://0.0.0.0:8004/
- Run through the following [QA steps](https://canonical-web-and-design.github.io/practices/workflow/qa-steps.html)
- Login as a publisher
- Go to http://0.0.0.0:8004/<your_snap>/builds and check to you can:
- Login with GitHub
- Try to link you snap with a github repo and check that the repos contains a mark when snapcraft.yaml is present
- Check if a snap with a github repo without snapcraft.yaml get the error

### CODE REVIEW

File: webapp/publisher/snaps/views.py [hunk 1]
```diff
@@ -1390,11 +1404,16 @@ def get_snap_builds(snap_name):
...
+        try:
+            context["github_user"] = github.get_username()
+            # Get snapcraft repositories sorted by snapcraft_yaml
+            context["github_repositories"] = sorted(
+                github.get_user_repositories(),
+                key=lambda k: k["snapcraft_yaml"],
+                reverse=True,
+            )
+        except Unauthorized:
+            pass
```
Comment (nottrobin) [inline webapp/publisher/snaps/views.py]: Why wouldn't we care about an unauthorized response?

Comment (jkfran) [reply webapp/publisher/snaps/views.py]: If we get an Unauthorized exception here that means the user is not logged in GitHub, that means github_user will be None in the template and we will show the login button.

Comment (nottrobin) [reply webapp/publisher/snaps/views.py]: Then maybe you could just put the line in here `context["github_user"] = None`? Just to show that we are deliberately doing something.

Comment (jkfran) [reply webapp/publisher/snaps/views.py]: Yes, it will make it easier to understand

File: webapp/api/github.py [hunk 2]
```diff
@@ -18,59 +18,121 @@ def __init__(self, access_token=None, session=api.requests.Session()):
...
+        gql = (
+            '''{
+            repository(name: "'''
+            + repo
+            + '''", owner: "'''
+            + owner
+            + """") {
+                id
+                nameWithOwner
+                yamlLocation1:
```
Comment (Lukewh) [inline webapp/api/github.py]: is there a way to template some of this out with GraphQL? For instance the repetition of these yamlLocations?

Comment (jkfran) [reply webapp/api/github.py]: Luke I changed this request to use the Rest API because we might don't need the user session to verify this.

**Comment by jkfran** [comment]
Hey @nottrobin , this is ready to be reviewed.

**Comment by nottrobin** [comment]
@jkfran think you could consider testing this branch a bit as @Lukewh suggests? I'd be happy to pair on that with you on Monday / Tuesday if you want?

**Comment by jkfran** [comment]
Hello @nottrobin ,

Thank you for your proposal, I have created a new commit in this branch with the changes for the API testing.

Let me know what do you think and if you think is better to pair for the review.

Thank you.

File: webapp/publisher/snaps/views.py [hunk 3]
```diff
@@ -1356,11 +1357,29 @@ def get_snap_builds(snap_name):
...
+    except Unauthorized:
+        context["github_user"] = None
+
     if lp_snap:
         # Git repository without GitHub hostname
         context["github_repository"] = lp_snap["git_repository_url"][19:]
+        github_owner, github_repo = context["github_repository"].split("/")
+
+        # Check if this repo has a snapcraft.yaml
+        context["yaml_file_exist"] = github.is_snapcraft_yaml_present(
```
Comment (nottrobin) [inline webapp/publisher/snaps/views.py]:
```suggestion
        context["yaml_file_exists"] = github.is_snapcraft_yaml_present(
```

File: webapp/publisher/snaps/views.py [hunk 4]
```diff
@@ -1356,11 +1357,29 @@ def get_snap_builds(snap_name):
     }
 
     # Get built snap in launchpad with this store name
+    github = GitHubAPI(flask.session.get("github_auth_secret"))
     lp_snap = launchpad.get_snap_by_store_name(details["snap_name"])
 
+    try:
+        context["github_user"] = github.get_username()
```
Comment (nottrobin) [inline webapp/publisher/snaps/views.py]: If this snap has `builds` lower down, do we need any information at all from their GitHub account? 'cos if not, we should move this request lower down in the logic so we're not making it if we don't need to.

Comment (jkfran) [reply webapp/publisher/snaps/views.py]: You are right, I moved it at the top when trying to use GraphQL to check the yaml file in the repo. But now that method doesn't need user authentication.

**Pull request merged**